brainiak/hyperparamopt/hpo.py

```diff
@@ -0,0 +1,231 @@
+    n = y.shape[0]
+    g = int(np.round(np.ceil(0.15 * n)))
+    ldata = z[0:g]
+    gdata = z[g:n]
+    lymin = ldata['y'].min()
+    lymax = ldata['y'].max()
+    weights = (lymax - ldata['y']) / (lymax - lymin)
+    lx = gmm_1d_distribution(ldata['x'], minlimit=minlimit,
+                             maxlimit=maxlimit, weights=weights)
+    gx = gmm_1d_distribution(gdata['x'], minlimit=minlimit, maxlimit=maxlimit)
+
+    samples = lx.get_samples(chains=10, points_per_chain=100)
+    ei = lx(samples) / gx(samples)
+
+    h = (x.max() - x.min()) / (10 * x.size)
+    # assumes prior of x is uniform -- should change for different priors
+    # d = np.abs(x - samples[ei.argmax()]).min()
+    # CDF(x+d/2) - CDF(x-d/2) < 1/(10*x.size) then reject else accept
+    s = 0
+    while (np.abs(x - samples[ei.argmax()]).min() < h):
+        ei[ei.argmax()] = 0
+        s = s + 1
+        if (s == samples.size):
+            break
+    xnext = samples[ei.argmax()]
+
+    return xnext
+
+
+def getSample(x, y, dist, minlimit=-np.inf, maxlimit=np.inf):
+    if (dist == 'GMM'):
+        return getNextSample(x, y, minlimit, maxlimit)
+    if (dist == 'uniform'):
+        return np.random.random() * (maxlimit - minlimit) + minlimit
+    if (dist == 'loguniform'):
+        return np.exp(np.random.random()
+                      * (np.log(maxlimit) - np.log(minlimit))
+                      + np.log(minlimit))
+
+
+def fmin(lossfn,
+         space,
+         maxevals,
+         trials,
+         init_random_evals=30,
+         explore_prob=0.2,
+         verbose=False):
+    """Find the minimum of function through hyper paramter optimization
+
+    Arguments
+    ---------
+
+    lossfn : function that takes in a dictionary and returns a real value
+             Function to be minimized
+
+    space : Dictionary specifying the range and distribution of
+            the hyperparamters
+
+    maxevals : int
+               Maximum number of evaluations of lossfn allowed
+
+    trials : list
+             Holds the output of the optimization trials
+             Need not be empty to begin with, new trials are appended
+             at the end
+
+    init_random_evals : int, default 30
+                        Number of random trials to initialize the
+                        optimization
+
+    explore_prob : double in [0, 1], default 0.2
+                   Controls the exploration-vs-exploitation ratio
+                   Currently 20% of trails are random samples
+
+    verbose : bool, default False
+              Get information on current point being processed
+
+    Returns
+    -------
+
+    best : trial entry (dictionary of hyperparameters)
+           Best hyperparameter setting found
+    """
+
+    for s in space:
+        if (space[s]['dist'] is not 'uniform' and
+                space[s]['dist'] is not 'loguniform'):
+            logger.error('Unsupported distribution for variable')
+            raise TypeError('Unknown distribution type for variable')
```

Comment:
This has been changed. space[s]['dict'] should now hold scipy.stats objects, so anything that does not support the "rvs()" method for generating random numbers throws a TypeError.